### PR TITLE
[SPARK-13033][ML][PySpark] Add import/export for ml.regression

### PR DESCRIPTION
**Pull request by Wenpei**
Add export/import for all estimators and transformers(which have Scala implementation) under pyspark/ml/regression.py.

@yanboliang Please help to review.
For doctest, I though it's enough to add one since it's common usage. But I can add to all if we want it.

### CODE REVIEW

File: python/pyspark/ml/regression.py [hunk 1]
```diff
@@ -447,7 +447,7 @@ def _create_model(self, java_model):
 
 
 @inherit_doc
-class DecisionTreeModel(JavaModel):
+class DecisionTreeModel(JavaModel, MLWritable, MLReadable):
     """Abstraction for Decision Tree models.
```
Comment (yanboliang) [inline python/pyspark/ml/regression.py]: @Wenpei Please check wether it supports ```save/load``` for the peer Scala implementation. Some algorithms such as ```DecisionTree``` did not support it currently. And you should add doc test that will test the correctness of your modification.

**Comment by Wenpei** [comment]
@yanboliang Sorry for last PR that I didn't check scala side.

For regression, there are only three algorithm support MLRead/MLWrite:
LinearRegression
IsotonicRegression
AFTSurvivalRegression

I add export/import api, and doc test currently.

But there is one issues here that doctest failed with below exception. It was caused we didn't set default value for "weightCol" (IsotonicRegression), "quantilesCol"(AFTSurvivalRegression) on scala code side. I add value when constructure to pass doctest, but I thought we should submit a jira for this. How about your idea?

Exception detail.
    ir2 = IsotonicRegression.load(ir_path)
Exception raised:
    Traceback (most recent call last):
      File "C:\Python27\lib\doctest.py", line 1289, in __run
        compileflags, 1) in test.globs
      File "<doctest __main__.IsotonicRegression[11]>", line 1, in <module>
        ir2 = IsotonicRegression.load(ir_path)
      File "C:\aWorkFolder\spark\spark-1.6.0-bin-hadoop2.6\spark-1.6.0-bin-hadoop2.6\python\lib\pyspark.zip\pyspark\ml\util.py", line 194, in load
        return cls.read().load(path)
      File "C:\aWorkFolder\spark\spark-1.6.0-bin-hadoop2.6\spark-1.6.0-bin-hadoop2.6\python\lib\pyspark.zip\pyspark\ml\util.py", line 148, in load
        instance._transfer_params_from_java()
      File "C:\aWorkFolder\spark\spark-1.6.0-bin-hadoop2.6\spark-1.6.0-bin-hadoop2.6\python\lib\pyspark.zip\pyspark\ml\wrapper.py", line 82, in _tran
fer_params_from_java
        value = _java2py(sc, self._java_obj.getOrDefault(java_param))
      File "C:\aWorkFolder\spark\spark-1.6.0-bin-hadoop2.6\spark-1.6.0-bin-hadoop2.6\python\lib\py4j-0.9-src.zip\py4j\java_gateway.py", line 813, in
_call__
        answer, self.gateway_client, self.target_id, self.name)
      File "C:\aWorkFolder\spark\spark-1.6.0-bin-hadoop2.6\spark-1.6.0-bin-hadoop2.6\python\lib\pyspark.zip\pyspark\sql\utils.py", line 45, in deco
        return f(*a, **kw)
      File "C:\aWorkFolder\spark\spark-1.6.0-bin-hadoop2.6\spark-1.6.0-bin-hadoop2.6\python\lib\py4j-0.9-src.zip\py4j\protocol.py", line 308, in get_
eturn_value
        format(target_id, ".", name), value)
    Py4JJavaError: An error occurred while calling o351.getOrDefault.
    : java.util.NoSuchElementException: Failed to find a default value for weightCol
        at org.apache.spark.ml.param.Params$$anonfun$getOrDefault$2.apply(params.scala:647)
        at org.apache.spark.ml.param.Params$$anonfun$getOrDefault$2.apply(params.scala:647)
        at scala.Option.getOrElse(Option.scala:120)
        at org.apache.spark.ml.param.Params$class.getOrDefault(params.scala:646)
        at org.apache.spark.ml.PipelineStage.getOrDefault(Pipeline.scala:43)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:483)
        at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:231)
        at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:381)
        at py4j.Gateway.invoke(Gateway.java:259)
        at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:133)
        at py4j.commands.CallCommand.execute(CallCommand.java:79)
        at py4j.GatewayConnection.run(GatewayConnection.java:209)
        at java.lang.Thread.run(Thread.java:745)

**Comment by yanboliang** [comment]
@Wenpei It looks like ```_transfer_params_from_java``` did not consider the params which do not have default value and we should handle them. Would you mind to create a jira to track this issue?

**Comment by Wenpei** [comment]
Sure, I will submit a jira, I thought we need fix it in scala side that ensure all parameter has default value.

**Comment by yanboliang** [comment]
It should not make all parameters have default value because of some params are not setting default value on purpose. I think we should modify ```_transfer_params_from_java``` to make it not to get the params which do not have default values.

**Comment by Wenpei** [comment]
Sure, good catch. I have submit a jira 13153 and submit a pr later

**Comment by Wenpei** [comment]
I need do a minor change for this PR after 13153 get resoved

**Comment by Wenpei** [comment]
@yanboliang  @mengxr 
Done for this PR.  Please help review and launch test.

File: python/pyspark/ml/regression.py [hunk 2]
```diff
@@ -179,6 +179,23 @@ class IsotonicRegression(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredicti
...
+    >>> path = tempfile.mkdtemp()
+    >>> ir_path = path + "/ir"
+    >>> ir.save(ir_path)
+    >>> ir2 = IsotonicRegression.load(ir_path)
+    >>> ir2.getIsotonic()
+    True
+    >>> model_path = path + "/ir_model"
+    >>> model.save(model_path)
+    >>> model2 = IsotonicRegressionModel.load(model_path)
+    >>> model.boundaries[0] == model2.boundaries[0]
```
Comment (yanboliang) [inline python/pyspark/ml/regression.py]: It's better to check ```model.boundaries == model2.boundaries```

File: python/pyspark/ml/regression.py [hunk 3]
```diff
@@ -697,6 +714,25 @@ class AFTSurvivalRegression(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredi
...
+    >>> path = tempfile.mkdtemp()
+    >>> aftsr_path = path + "/aftsr"
+    >>> aftsr.save(aftsr_path)
+    >>> aftsr2 = AFTSurvivalRegression.load(aftsr_path)
+    >>> aftsr2.getMaxIter()
+    100
+    >>> model_path = path + "/aftsr_model"
+    >>> model.save(model_path)
+    >>> model2 = AFTSurvivalRegressionModel.load(model_path)
+    >>> model.coefficients[0] == model2.coefficients[0]
```
Comment (yanboliang) [inline python/pyspark/ml/regression.py]: Ditto, because the coefficients is 2-dimension vector in this test case.

**Comment by yanboliang** [comment]
@Wenpei Please pay attention to the status of #11197 and update this PR corresponding when it get merged.

**Comment by Wenpei** [comment]
@yanboliang OK.

**Comment by Wenpei** [comment]
@yanboliang I complete this pr, please take a look

File: python/pyspark/ml/regression.py [hunk 4]
```diff
@@ -172,6 +172,16 @@ class IsotonicRegression(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredicti
...
+    >>> ir_path = temp_path + "/ir"
+    >>> ir.save(ir_path)
+    >>> ir2 = IsotonicRegression.load(ir_path)
+    >>> ir2.getIsotonic()
+    True
+    >>> model_path = temp_path + "/ir_model"
+    >>> model.save(model_path)
+    >>> model2 = IsotonicRegressionModel.load(model_path)
+    >>> model.boundaries == model2.boundaries
+    True
```
Comment (yanboliang) [inline python/pyspark/ml/regression.py]: We should also check equality of ```predictions```.

File: python/pyspark/ml/regression.py [hunk 5]
```diff
@@ -690,6 +700,18 @@ class AFTSurvivalRegression(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredi
...
+    >>> aftsr2 = AFTSurvivalRegression.load(aftsr_path)
+    >>> aftsr2.getMaxIter()
+    100
+    >>> model_path = temp_path + "/aftsr_model"
+    >>> model.save(model_path)
+    >>> model2 = AFTSurvivalRegressionModel.load(model_path)
+    >>> model.coefficients == model2.coefficients
+    True
+    >>> model.intercept == model2.intercept
+    True
```
Comment (yanboliang) [inline python/pyspark/ml/regression.py]: We should also check equality of model ```scale```.

**Comment by yanboliang** [comment]
Looks good except minor issues.

**Comment by yanboliang** [comment]
Jenkins, test this please.

**Comment by Wenpei** [comment]
@mengxr @srowen Can you add me to white list ? or help to launch a jenkins test for this?

**Comment by srowen** [comment]
Jenkins, test this please.

**Comment by yanboliang** [comment]
LGTM

**Comment by yanboliang** [comment]
LGTM

**Comment by yanboliang** [comment]
LGTM

**Comment by yanboliang** [comment]
LGTM

**Comment by mengxr** [comment]
Merged into master. Thanks!

**Pull request closed without merging**